### PR TITLE
More-flexible user-specified dependency constraints

### PR DESCRIPTION
**Pull request by scheibelp**
See: https://github.com/spack/spack/issues/8133

There are many cases where Spack's greedy concretization algorithm makes choices that result in conflicts that it cannot undo. For example (from the linked issue):

```
$ spack spec ecp-proxy-apps
Input spec
--------------------------------
ecp-proxy-apps

Concretized
--------------------------------
==> Error: Multiple providers found for 'jpeg': ['libjpeg', 'libjpeg-turbo@1.5.90%gcc@5.4.0 arch=linux-gentoo2-x86_64']
```

This is partly due to the fact that the concretizer algorithm runs in passes: each pass (a) expands known dependencies and (b) chooses variant settings (since (b) affects (a), this happens in a loop). It so happens in this case that an earlier pass chose `libjpeg-turbo` as the implementation of the `jpeg` virtual, but in a later pass found that some dependency specifically wanted the `libjpeg` implementation.

There are many incarnations of this issue. The typical first response is to give a hint to the concretizer:

`spack spec ecp-proxy-apps ^libjpeg`

which can often result in an error like the following:

```
==> Error: ecp-proxy-apps does not depend on libjpeg
```

This is because the concretization algorithm contains an intended-to-be-useful guard which checks whether user-specified constraints actually apply to the DAG. The problem is that this guard *also* executes on every pass and doesn't account for the fact that user-specified constraints might only apply to dependencies that show up in later passes.

SO... what does this PR do? It relaxes the guard and only applies it at the end, so you can do

`spack speck ecp-proxy-apps ^libjpeg ^opengl`

and it won't complain (because those dependencies eventually show up in the DAG).

### CODE REVIEW

File: lib/spack/spack/spec.py [hunk 1]
```diff
@@ -2221,24 +2234,23 @@ def normalize(self, force=False, tests=False):
         # Get all the dependencies into one DependencyMap
         spec_deps = self.flat_dependencies(copy=False)
 
+        full_spec_deps = dict(spec_deps)
+        if user_spec_deps:
+            for name, spec in user_spec_deps.items():
+                if name not in full_spec_deps:
```
Comment (tgamblin) [inline lib/spack/spack/spec.py]: Won't this skip user constraints whose name matches a package constraint?  I think you want to add:

```python
    ...
else:
    full_spec_deps[name].constrain(spec)
```

File: lib/spack/spack/spec.py [hunk 2]
```diff
@@ -2221,24 +2234,23 @@ def normalize(self, force=False, tests=False):
         # Get all the dependencies into one DependencyMap
         spec_deps = self.flat_dependencies(copy=False)
 
+        full_spec_deps = dict(spec_deps)
```
Comment (tgamblin) [inline lib/spack/spack/spec.py]: Why create a new `dict`?  At this point you should just update `spec_deps` and not bother with another `dict`.  It probably needs another name now that it includes more.

**Comment by gfursin** [comment]
Hi! I had a similar issue when trying to install mxnet via spack (https://groups.google.com/forum/#!topic/spack/wKcAathCgVo) and this patch worked for me. Thanks!

**Comment by citibeth** [comment]
> This is because the concretization algorithm contains an intended-to-be-useful guard which checks whether user-specified constraints actually apply to the DAG.

@tgamblin  @scheibelp Why do we have this guard at all?  Why have we created a system whereby constraints specified on the command line have subtly different semantics than constraints specified in `packages.yaml`.  For simplicity sake and predictability if nothing else, I would prefer they both work the same.  I don't see how the guard is useful at all.

**Pull request merged**